chapters/04-functional.md

```diff
@@ -0,0 +1,105 @@
+these feature specifications are usually written by the product owner in
+collaboration with the development team. The developers would then implement
+them. After it's implemented, Cucumber would run the specifications to ensure
+that the specifications are actually working as intended. In some sense, the
+specifications are now the tests and they can be run as part of the CI
+pipeline.
+
+In summary a BDD cycle would look like this:
+- Product owner writes feature specifications in collaboration with the development team.
+- Developers implement the feature.
+- Tester or developer runs the Cucumber tests to ensure specifications were built.
+
+## Example of a feature specificaton
+
+In Cucumber, product specifications are written in features files, these
+feature files uses a Domain Specific Language (DSL) known as Gherkin. It uses a
+Given-When-Then language that is close to natural language. Here's an example
+of a feature file in Cucumber that specifies a feature for a calculator
+program:
+
+```
+Feature: Addition
+  Scenario: Calculate the sum of two integers
+    Given two integers
+    When the add button is pressed
+    Then the sum of the two integers should be displayed
+```
+
+Being very close to a natural language, it is readable by stakeholders in the
+organization and non-technical product owners. In the above example it is clear
+how the user is interacting with the program, what the inputs and preconditions
+are, and what should be expected. These properties are described using the
+Given-When-Then language:
+
+- `Given`: describes the preconditions of the feature
+- `When`: describes the action that the user carries out in the feature
+- `Then`: desribes what's expected after the action is performed
+
+In Cucumber, these are known as steps, each step when evaluated has a
+predefined behavior. Although Cucumber already come with some predefined steps,
+developers and testers can extend the language to suit their application. For
+example, the following step uses the `click_on` method defined in the
+`capybara` ruby gem. It navigates to the home page when Cucumber sees a step
+that says `When the home button is clicked`:
+
+```
+When(/^the home button is clicked$/) do
+  click_on 'Home'
+end
+```
+
+## Executable user stories
+
+In an agile team, this feature file is also known as a user story. It defines a
+user's requirements in the way the user would interact with the software. Apart
+from being just a document to communicate the requirements of a user, this
+feature file can be executed and ran against the product to ensure that the
+product meets the user's requirement and use case as described in the story. To
+do so, Cucumber evaluates the Gherkin language, populates the test environment
+with the preconditions and input, and then simulates the interaction with the
+program, and lastly it checks the actual behavior of the program against the
+desired behavior as specified in the feature file.
+
+## Closing thoughts
```

Comment:
Could we add something here on how _we_ use cucumber (via aruba)? Since the book is about our software testing experience. Could be just one of our functional tests + explanation on the feature it's testing.